packages/grafana-ui/src/components/Combobox/Combobox.tsx

```diff
@@ -167,20 +169,54 @@ export const Combobox = <T extends string | number>(props: ComboboxProps<T>) =>
     return typeof valueProp === 'object' ? valueProp : { value: valueProp, label: valueProp.toString() };
   }, [selectedItemIndex, isAsync, valueProp, allOptions]);
 
-  const menuId = `downshift-${useId().replace(/:/g, '--')}-menu`;
-  const labelId = `downshift-${useId().replace(/:/g, '--')}-label`;
+  const menuId = `${baseId}-downshift-menu`;
+  const labelId = `${baseId}-downshift-label`;
 
   const styles = useStyles2(getComboboxStyles);
 
-  const virtualizerOptions = {
+  // Injects the group header for the first rendered item into the range to render.
+  // Accepts the range that useVirtualizer wants to render, and then returns indexes
+  // to actually render.
+  const rangeExtractor = useCallback(
+    (range: Range) => {
+      const startIndex = Math.max(0, range.startIndex - range.overscan);
+      const endIndex = Math.min(filteredOptions.length - 1, range.endIndex + range.overscan);
+      const rangeToReturn = Array.from({ length: endIndex - startIndex + 1 }, (_, i) => startIndex + i);
+
+      // If the first item doesn't have a group, no need to find a header for it
+      if (filteredOptions[rangeToReturn[0]]?.group) {
+        // Find the first group header that's not in view
+        for (let offscreenIndex = startIndex - 1; offscreenIndex >= 0; offscreenIndex--) {
+          if (isNewGroup(filteredOptions[offscreenIndex], filteredOptions[offscreenIndex - 1])) {
+            rangeToReturn.unshift(offscreenIndex);
+            break;
+          }
+        }
+      }
+
+      return rangeToReturn;
```

Comment:
Isn't this basically a linear search, which become a performance bottleneck when there's many items.

Reply:
Yeah, this is probably pretty linear, but most of the time its only operating on a smaller subset of options - It starts at the first visible item in the menu (depending on scroll position), and backtracks until it finds the beginning of the current group.

The worst case scenario would be if all the options are in one exceptionally large group and the user has scrolled to the bottom.

Because the comparison here isn't very expensive (it's just comparing one string to another), I don't think this'll be too much of an issue in practice.

Thoughts? Is there another approach I'm missing?

Reply:
Maybe saving a map of the index of known group titles would be an option? Then you'd avoid having to redo this on pretty much every scroll.